ddl/ttl.go

```diff
@@ -0,0 +1,133 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ddl
+
+import (
+	"fmt"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/meta"
+	"github.com/pingcap/tidb/parser"
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/dbterror"
+)
+
+func onTTLInfoRemove(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error) {
+	tblInfo, err := GetTableInfoAndCancelFaultJob(t, job, job.SchemaID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	tblInfo.TTLInfo = nil
+	ver, err = updateVersionAndTableInfo(d, t, job, tblInfo, true)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	job.FinishTableJob(model.JobStateDone, model.StatePublic, ver, tblInfo)
+	return ver, nil
+}
+
+func onTTLInfoChange(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error) {
+	var ttlInfo model.TTLInfo
+	if err := job.DecodeArgs(&ttlInfo); err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	tblInfo, err := GetTableInfoAndCancelFaultJob(t, job, job.SchemaID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	tblInfo.TTLInfo = &ttlInfo
+	ver, err = updateVersionAndTableInfo(d, t, job, tblInfo, true)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	job.FinishTableJob(model.JobStateDone, model.StatePublic, ver, tblInfo)
+	return ver, nil
+}
+
+func onTTLEnableChange(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error) {
+	var ttlEnable bool
+	if err := job.DecodeArgs(&ttlEnable); err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	tblInfo, err := GetTableInfoAndCancelFaultJob(t, job, job.SchemaID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	tblInfo.TTLEnable = &ttlEnable
+	ver, err = updateVersionAndTableInfo(d, t, job, tblInfo, true)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	job.FinishTableJob(model.JobStateDone, model.StatePublic, ver, tblInfo)
+	return ver, nil
+}
+
+func checkTTLInfoValid(ctx sessionctx.Context, tblInfo *model.TableInfo) error {
+	if err := checkTTLIntervalExpr(ctx, tblInfo.TTLInfo); err != nil {
+		return err
+	}
+
+	return checkTTLInfoColumnType(tblInfo)
+}
+
+func checkTTLIntervalExpr(ctx sessionctx.Context, ttlInfo *model.TTLInfo) error {
+	// FIXME: use a better way to validate the interval expression in ttl
+	// TODO: check whether the expression is a constant, though it can pass this function, it still may contain a variable
```

Comment:
Could we just only allow `ValueExpr` for safe now? I think it is enough for the first version. Maybe we can open more expressions if needed later.

Reply:
Yes. The `ValueExpr` is much better.

The `DEFAULT` expression uses a `DefaultValueExpr`, which includes signed literal, and some function calls like `NOW()`. I think it's fine to just support literal in this situation :+1: .

I have modified the parser accordingly in https://github.com/pingcap/tidb/pull/39341.